### ISSUE TITLE
Cannot Reproduce the MOT17 Scores

### ISSUE THREAD

**Issue by emirhanbayar** [mikel-brostrom/yolo_tracking#1156]
### Search before asking

- [X] I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.

### Question

Hello,

I encounter an inconsistency between your implementation's HOTA scores and those stated in the paper of StrongSORT.

Setup:
Tracker: StrongSORT (did not change any parameter)
Detections: from https://drive.google.com/drive/folders/1Zk6TaSJPbpnqbz1w4kfhkKFCEzQbjfp_?usp=sharing
Re-ID: osnet_x0_25_msmt17.pt

HOTA scores after running your implementation:
MOT-17-Train: 63.8
MOT-17-Test: 59.69

HOTA scores claimed in the original paper:
MOT-17-Train: 68.2 (StrongSORT v4 in Table 1)
MOT-17-Test: 63.5 (StrongSORT online in Table 4)

I suspected this results from using a different Re-ID and switched to using features provided by the authors of StrongSORT. However, this even reduced the HOTA score on the MOT-17-Train.

Do you have any idea why this is happening? I would appreciate any comment on this.

Thanks in advance.

**Comment by mikel-brostrom** [comment]
Thank you for sharing these experiment results @emirhanbayar 

> Do you have any idea why this is happening?

I have not checked the config parameters for StrongSORT in a long time, I suggest you to start there

> Re-ID: osnet_x0_25_msmt17.pt

According to my OSNet experiments `osnetx1_0_dukemtcereid.pt` should perform the best on MOT17.

**Comment by luke110** [comment]
> ### 询问前先搜索
> * [x] 我搜索了 Yolo Tracking[问题](https://github.com/mikel-brostrom/yolo_tracking/issues)，没有发现类似的错误报告。
> 
> ### 问题
> 你好，
> 
> 我发现您的实现的 HOTA 分数与 StrongSORT 论文中所述的分数不一致。
> 
> 设置： 跟踪器：StrongSORT（未更改任何参数） 检测：来自https://drive.google.com/drive/folders/1Zk6TaSJPbpnqbz1w4kfhkKFCEzQbjfp_?usp=sharing 重新 ID：osnet_x0_25_msmt17.pt
> 
> 运行实施后的 HOTA 分数： MOT-17-Train：63.8 MOT-17-Test：59.69
> 
> 原论文中声称的 HOTA 分数： MOT-17-Train：68.2（表 1 中的 StrongSORT v4） MOT-17-Test：63.5（表 4 中的 StrongSORT online）
> 
> 我怀疑这是由于使用不同的 Re-ID 造成的，因此转而使用 StrongSORT 作者提供的功能。然而，这甚至降低了 MOT-17-Train 的 HOTA 分数。
> 
> 您知道为什么会发生这种情况吗？如果对此有任何评论，我将不胜感激。
> 
> 提前致谢。



> ### Search before asking
> * [x]  I have searched the Yolo Tracking [issues](https://github.com/mikel-brostrom/yolo_tracking/issues) and found no similar bug report.
> 
> ### Question
> Hello,
> 
> I encounter an inconsistency between your implementation's HOTA scores and those stated in the paper of StrongSORT.
> 
> Setup: Tracker: StrongSORT (did not change any parameter) Detections: from https://drive.google.com/drive/folders/1Zk6TaSJPbpnqbz1w4kfhkKFCEzQbjfp_?usp=sharing Re-ID: osnet_x0_25_msmt17.pt
> 
> HOTA scores after running your implementation: MOT-17-Train: 63.8 MOT-17-Test: 59.69
> 
> HOTA scores claimed in the original paper: MOT-17-Train: 68.2 (StrongSORT v4 in Table 1) MOT-17-Test: 63.5 (StrongSORT online in Table 4)
> 
> I suspected this results from using a different Re-ID and switched to using features provided by the authors of StrongSORT. However, this even reduced the HOTA score on the MOT-17-Train.
> 
> Do you have any idea why this is happening? I would appreciate any comment on this.
> 
> Thanks in advance.

How did you train to get your own weight file?

**Comment by emirhanbayar** [comment]
I did not train anything. If you mean the reID weight file (osnet_x0_25_msmt17.pt). I chose it from the ones that are listed at: https://github.com/mikel-brostrom/yolo_tracking/blob/master/boxmot/appearance/reid_model_factory.py

**Comment by mikel-brostrom** [comment]
First of all. Which detector are you using?

**Comment by mikel-brostrom** [comment]
> MOT17 train detections from: https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr MOT17 test detections from: https://drive.google.com/drive/folders/16oC71eiWhDe5wJaYOyWkgE3h-zq2WGgi
> 
> The authors say that they used YOLOX.

Ok. Then you are using the right detections :smile:.

- What MOTA and IDF1 scores do you get?
- Have you fine-tuned the parameters?

**Comment by emirhanbayar** [comment]
> * What MOTA and IDF1 scores do you get?

I had recorded the following results:
MOT17 train: HOTA: 68.3, MOTA: 66.619, IDF1: 73.868
MOT17 test: HOTA: 59.69, MOTA: 71.54, IDF1: 73.63

I newly ran the code with the following parameters:
model_weights: osnet_x0_25_msmt17.pt
max_dist: 0.2
max_iou_dist: 0.7
max_age: 30
n_init: 1
nn_budget: 100
mc_lambda: 0.995
ema_alpha: 0.9

And got the following result:
MOT17 train: HOTA: 64.033, MOTA: 66.365, IDF1: 74.329



> * Have you fine-tuned the parameters?
No.

**Comment by mikel-brostrom** [comment]
Have you tried:

https://drive.google.com/drive/folders/1wdG-vJOMGynf5QjqEa1jNZNYtz0PQqPu

? 

Check this: https://github.com/dyhBUPT/StrongSORT/issues/74

**Comment by emirhanbayar** [comment]
> https://drive.google.com/drive/folders/1wdG-vJOMGynf5QjqEa1jNZNYtz0PQqPu

I had already modified the ground truth accordingly. Now, I downloaded them from the link that you shared. The result is exactly the same.

**An update:** I fed both detections and features that are shared by the authors of StrongSORT as input. I got:

HOTA: 63.105
MOTA: 66.354
IDF1: 71.77

I am not sure. Maybe, I am making a mistake. You can inspect how I modified your implementation by looking at my fork at: https://github.com/emirhanbayar/yolo_tracking

To run the code:

1. Place MOT17 into the folder named data 
2. Remove all the sequences in the MOT17 that end with -DPM or -SDP. Only the ones that end with -FRCNN should remain.
3. Download detections from https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr
4. Place detections into the folder named MOT17_train_dets
5. Run python run.py --tracker strongsort 
6. The results will be saved into the folder named results

**Comment by mikel-brostrom** [comment]
Great! We are getting closer... :smile:

| MOT-17 | hota |
| ------------- | ------------- |
| paper implementation | 68.2 |
| boxmot | 65.353 |

You can now try `osnetx1_0_dukemtcereid.pt`. It is the best lightweight performant model for MOT17.

**Comment by emirhanbayar** [comment]
> Notice you won't get the paper result as we are not using the exact same ReID model.

Actually, hota 65.4 is obtained by feeding reid features from the ones that pre-extracted and shared by authors of StrongSORT. Neither detector nor thr reID model is used from boxmot.

**Comment by mikel-brostrom** [comment]
I am unfamiliar with this data

> MOT17 train detections from: https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr MOT17 test detections from: https://drive.google.com/drive/folders/16oC71eiWhDe5wJaYOyWkgE3h-zq2WGgi
> 
> The authors say that they used YOLOX.

Are the features normalized?

**Comment by emirhanbayar** [comment]
Sorry for the late reply. The entries are generally between -2 and 4. Some examples of distributions of the feature vectors from the detections from MOT17 sequence 2:

![135](https://github.com/mikel-brostrom/yolo_tracking/assets/85796946/c68bc899-425e-4435-af60-1a6f6a86d2c5)

![152](https://github.com/mikel-brostrom/yolo_tracking/assets/85796946/77cdf56e-6874-4d3c-8fce-af175d1dba54)

You can see the distributions for each of the detections at: https://drive.google.com/file/d/1Ef8zms1Ph5JOn5kRjpRYc7kdYHIe3lCw/view?usp=sharing

**Comment by mikel-brostrom** [comment]
I just noticed that the MOTA score (66.354) you get is quite low compared to the reported one (77.0). This could be due to some misconfiguration in the Kalman Fillter. You could try importing the official one and that one instead.

**Comment by mikel-brostrom** [comment]
Aaand you need to normalize the embeddings for better IDF1 results :smile:

**Comment by lvfengkun** [comment]
Hi,guys,I have partially addressed your issue by setting "track.time_since_update" of "strong_sort" to >2, resulting in an HOTA of 68.184 on MOT17. The deviation from the results in the paper is less than 1, indicating that there is still room for optimization. The reason why adjusting "time_since_update" proves to be effective in enhancing HOTA is currently under exploration. If you have any findings on this matter, I would appreciate the opportunity for further discussion.

**Comment by lvfengkun** [comment]
> > https://drive.google.com/drive/folders/1wdG-vJOMGynf5QjqEa1jNZNYtz0PQqPu
> 
> 我已经相应地修改了基本事实。现在，我从您共享的链接下载了它们。结果完全相同。
> 
> **更新：**我将 StrongSORT 作者共享的检测和功能作为输入提供，并保留了与您的实现中相同的参数。在 MOT17 列车上，我得到了：
> 
> HOTA：63.105 MOTA：66.354 IDF1：71.77
> 
> 更改了 StrongSORT 原始实现中的参数：max_dist=0.4、max_iou_dist=0.7、max_age=30、n_init=1、nn_budget=100、mc_lambda=0.98、ema_alpha=0.9、
> 
> 然后，在 MOT17 列车上，我得到了：
> 
> HOTA：65.353 MOTA：66.758 IDF1：76.239
> 
> 我不确定。也许，我犯了一个错误。您可以通过查看我的 fork 来检查我如何修改您的实现： https://github.com/emirhanbayar/yolo_tracking
> 
> 要运行代码，请执行以下操作：
> 
> 1. 将 MOT17 放入名为 data 的文件夹中
> 2. 删除 MOT17 中以 -DPM 或 -SDP 结尾的所有序列。只有以 -FRCNN 结尾的那些应该保留。
> 3. 从 https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr 下载检测
> 4. 将检测放入名为 MOT17_train_dets 的文件夹中
> 5. 运行 python run.py --tracker strongsort
> 6. 结果将保存到名为 results 的文件夹中



> > https://drive.google.com/drive/folders/1wdG-vJOMGynf5QjqEa1jNZNYtz0PQqPu
> 
> I had already modified the ground truth accordingly. Now, I downloaded them from the link that you shared. The result is exactly the same.
> 
> **An update:** I fed both detections and features that are shared by the authors of StrongSORT as input and kept the same parameters as in your implementation. On the MOT17 train, I got:
> 
> HOTA: 63.105 MOTA: 66.354 IDF1: 71.77
> 
> Changed parameters as in the original implemenataion of StrongSORT: max_dist=0.4, max_iou_dist=0.7, max_age=30, n_init=1, nn_budget=100, mc_lambda=0.98, ema_alpha=0.9,
> 
> Then, on the MOT17 train, I got:
> 
> HOTA: 65.353 MOTA: 66.758 IDF1: 76.239
> 
> I am not sure. Maybe, I am making a mistake. You can inspect how I modified your implementation by looking at my fork at: https://github.com/emirhanbayar/yolo_tracking
> 
> To run the code:
> 
> 1. Place MOT17 into the folder named data
> 2. Remove all the sequences in the MOT17 that end with -DPM or -SDP. Only the ones that end with -FRCNN should remain.
> 3. Download detections from https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr
> 4. Place detections into the folder named MOT17_train_dets
> 5. Run python run.py --tracker strongsort
> 6. The results will be saved into the folder named results

You have to comment out the part from if tlwh[0] < 0: to tlwh[1] = 0, I'm not sure why you want to do this, but no doubt this reduces HOTA

**Comment by lvfengkun** [comment]
> > https://drive.google.com/drive/folders/1wdG-vJOMGynf5QjqEa1jNZNYtz0PQqPu
> 
> I had already modified the ground truth accordingly. Now, I downloaded them from the link that you shared. The result is exactly the same.
> 
> **An update:** I fed both detections and features that are shared by the authors of StrongSORT as input and kept the same parameters as in your implementation. On the MOT17 train, I got:
> 
> HOTA: 63.105 MOTA: 66.354 IDF1: 71.77
> 
> Changed parameters as in the original implemenataion of StrongSORT: max_dist=0.4, max_iou_dist=0.7, max_age=30, n_init=1, nn_budget=100, mc_lambda=0.98, ema_alpha=0.9,
> 
> Then, on the MOT17 train, I got:
> 
> HOTA: 65.353 MOTA: 66.758 IDF1: 76.239
> 
> I am not sure. Maybe, I am making a mistake. You can inspect how I modified your implementation by looking at my fork at: https://github.com/emirhanbayar/yolo_tracking
> 
> To run the code:
> 
> 1. Place MOT17 into the folder named data
> 2. Remove all the sequences in the MOT17 that end with -DPM or -SDP. Only the ones that end with -FRCNN should remain.
> 3. Download detections from https://drive.google.com/drive/folders/1zzzUROXYXt8NjxO1WUcwSzqD-nn7rPNr
> 4. Place detections into the folder named MOT17_train_dets
> 5. Run python run.py --tracker strongsort
> 6. The results will be saved into the folder named results

You have to comment out the part from if tlwh[0] < 0: to tlwh[1] = 0, I'm not sure why you want to do this, but no doubt this reduces HOTA

**Comment by lvfengkun** [comment]
> > Hi,guys,I have partially addressed your issue by setting "track.time_since_update" of "strong_sort" to >2, resulting in an HOTA of 68.184 on MOT17. The deviation from the results in the paper is less than 1, indicating that there is still room for optimization. The reason why adjusting "time_since_update" proves to be effective in enhancing HOTA is currently under exploration. If you have any findings on this matter, I would appreciate the opportunity for further discussion.
> 
> https://github.com/mikel-brostrom/yolo_tracking/blob/eecf37a22175c20526e70e30d8f0dfe8a2c96a09/boxmot/trackers/strongsort/strong_sort.py#L84C1-L84C73

yeah, that's it, change >= to >

**Comment by mikel-brostrom** [comment]
I believe this will cause to output the last detected bbox for 2 frames more. So when a track fails in associating with a detection, the last detected bbox will be output'ed anyways instead of not giving an output at all. Could be beneficial for when the track is lost for a few frames.

**Comment by lvfengkun** [comment]
Your intuition is correct that increasing this number will make the output more tracks, but in fact, if you look at strongsort's results, you'll see that the optimal HOTA output txt file size is larger than the boxmot output txt file size, in other words, strongsort's optimal result produces more and longer tracks. In my tests, setting this number to 1 to 3 would produce the same HOTA, 68.184, but the output trajectory would be different, i.e., the larger the number, the more the trajectory. I think understanding the meaning of this number would help improve strongsort and other trackers in boxmot



---Original---
From: ***@***.***&gt;
Date: Fri, Dec 15, 2023 22:32 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [mikel-brostrom/yolo_tracking] Cannot Reproduce the MOT17 Scores(Issue #1156)




 
I believe this will cause to output the last detected bbox for 2 frames more. So when a track fails in associating with a detection, the last detected bbox will be output'ed anyways instead of not giving an output at all. Could be beneficial for when the track is lost for a few frames.

**Comment by lvfengkun** [comment]
In my tests, the optimal result for this number should be 1 to 3 (&gt; instead of &gt; =), and the meaning of this number should be the maximum number of frames the track has not been updated. Now, I think we should look at other parameters to determine why boxmot's HOTA is less than the paper value, after all, it would be very strange to use strongsort's detection and features directly to produce different results.&nbsp;
ps: Instead of ecc, I saved the ecc results from boxmot and ran them in strongsort, producing similar results



---Original---
From: ***@***.***&gt;
Date: Fri, Dec 15, 2023 22:32 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [mikel-brostrom/yolo_tracking] Cannot Reproduce the MOT17 Scores(Issue #1156)




 
I believe this will cause to output the last detected bbox for 2 frames more. So when a track fails in associating with a detection, the last detected bbox will be output'ed anyways instead of not giving an output at all. Could be beneficial for when the track is lost for a few frames.

**Comment by mikel-brostrom** [comment]
> Hi,guys,I have partially addressed your issue by setting "track.time_since_update" of "strong_sort" to >2, resulting in an HOTA of 68.184 on MOT17. The deviation from the results in the paper is less than 1, indicating that there is still room for optimization. The reason why adjusting "time_since_update" proves to be effective in enhancing HOTA is currently under exploration. If you have any findings on this matter, I would appreciate the opportunity for further discussion.

I believe the remaining performance increase can be easily be obtained by fine-tuning.

**Comment by mikel-brostrom** [comment]
I have also noticed that modifying `track.time_since_update >= 1` leads to errors due to the detection filtering based on tracking results:

https://github.com/ultralytics/ultralytics/blob/38eaf5e29f2a269ee50de659470c91ae44bb23f8/ultralytics/trackers/track.py#L64-L66

For making predictions ahead some modifications are needed
